confgenerator/feature.go

```diff
@@ -0,0 +1,176 @@
+package confgenerator
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+var ErrTrackingInlineStruct = errors.New("cannot have tracking on inline struct")
+
+type Feature struct {
+	Module, Kind, Type, Key, Value string
+}
+
+func ExtractFeatures(uc *UnifiedConfig) ([]Feature, error) {
+	allFeatures := getOverriddenDefaultPipelines(uc)
+
+	if uc.Metrics != nil {
+		for _, r := range uc.Metrics.Receivers {
+			features, err := trackingFeatures(r, MetricsReceiverTypes.Subagent, MetricsReceiverTypes.Kind, r.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+		for _, p := range uc.Metrics.Processors {
+			features, err := trackingFeatures(p, MetricsProcessorTypes.Subagent, MetricsProcessorTypes.Kind, p.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+	}
+	if uc.Logging != nil {
+		for _, r := range uc.Logging.Receivers {
+			features, err := trackingFeatures(r, LoggingReceiverTypes.Subagent, LoggingReceiverTypes.Kind, r.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+		for _, p := range uc.Logging.Processors {
+			features, err := trackingFeatures(p, LoggingProcessorTypes.Subagent, LoggingProcessorTypes.Kind, p.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+	}
+
+	return allFeatures, nil
+}
+
+func trackingFeatures(component interface{}, module, kind, typ string) ([]Feature, error) {
+	return trackingFeaturesHelper(component, module, kind, typ, "")
+}
+
+func trackingFeaturesHelper(component any, module, kind, typ, prefix string) ([]Feature, error) {
+	if component == nil {
+		return []Feature{}, nil
+	}
+	c := reflect.ValueOf(component)
+	t := c.Type()
+
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	var features []Feature
+	// Iterate over all available fields and read the tag value
+	for i := 0; i < t.NumField(); i++ {
+		// Get the field, returns https://golang.org/pkg/reflect/#StructField
+		field := t.Field(i)
+
+		// Grab actual value instead of pointer
+		v := reflect.Indirect(reflect.Indirect(c).FieldByName(field.Name))
+		if v.IsZero() {
+			continue
+		}
+
+		// Get the field tag value
+		trackingTag := field.Tag.Get("tracking")
+		tags := strings.Split(trackingTag, ",")
+		trackingName := tags[0]
+
+		// Check to see if value needs to be overridden
+		var overridden string
+		if len(tags) > 1 {
+			overridden = tags[1]
+		}
+		yamlTag := field.Tag.Get("yaml")
+		isInline := strings.Contains(yamlTag, ",inline")
```

Comment:
Why bother handling some inline fields specially? Why not just expose them all? That will still produce a predictable name, which is all that matters for our debugging.

(I'm worried about trying to parse a different package's struct tag like this, because we might not have the same parsing behavior as they do.)

Reply:
Inline structs are omitted because no name is included.

The approach for parsing the yaml tag is: https://github.com/goccy/go-yaml/blob/ecececdd257a265811b1646bc8e96b85cc2d2b0d/struct.go#L42-L79

Reply:
Sure, but since this is for internal consumption, you don't need to care about the YAML tag. Embedded struct fields still have names from a reflection standpoint.

Reply:
Sync'd offline to understand the context and tradeoffs of the options. To summarize them:

A) Completely generic code walks the config struct and exposes every tagged variable with its struct path and value
* Pros
  * Simplest to implement
* Cons
  * The feature tracking metrics are not organized by features (Features generally map to receivers or processors). The path is flattened. More painful to query.

B) Non-generic code explicitly looks at top-level receiver and processor settings, and exposes the tagged ones with their struct field name and value
* Pros
  * Medium for implementation complexity
  * The feature tracking metrics are organized by features (Features generally map to receivers or processors).
* Cons
  * If the internal config parameter struct hierarchy or naming changes in the future, the corresponding feature tracking metrics labels will change, breaking existing feature tracking charts / dashboards unless someone notices and modify the charts / dashboards accordingly. The mitigation to prevent such changes is to stay locked in with the current structure naming and hierarchy.

C) Non-generic code explicitly looks at top-level receiver and processor settings, and exposes the tagged ones with an inferred YAML field name and value
* Pros
  * Internal config parameter struct hierarchy or naming can change freely. Because there is an extra layer to map to the yaml field names (aka user visible config parameter names that won't change without a breaking major release), the ingested feature tracking metrics can stay the same.
  * The feature tracking metrics are organized by features (Features generally map to receivers or processors).
* Cons
  * Most complex to implement

Given the tradeoffs, the decision is to stick with option C, aka what is implemented currently.